modules/sc-mesh-secure-deployment/src/2_0/features/cbma/auth/authClient.py

```diff
@@ -0,0 +1,127 @@
+        self.interface = interface
+        self.secure_client_socket = None
+        self.logger = logger_instance.get_logger()
+        self.ca = ca_path
+        self.mymac = get_mac_addr(self.interface)
+        self.server_mac = server_mac
+        self.mua = mua
+
+    def establish_connection(self):
+        # Create an SSL context
+        context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
+        context.verify_mode = ssl.CERT_REQUIRED
+
+        # Uncomment to enable Certificate Revocation List (CRL) check
+        # context.verify_flags = ssl.VERIFY_CRL_CHECK_LEAF
+
+        context.load_verify_locations(glob.glob(self.ca)[0])
+        context.load_cert_chain(
+            certfile=glob.glob(f"{self.CERT_PATH}/macsec_{self.mymac.replace(':', '')}.crt")[0],
+            keyfile=glob.glob(f"{self.CERT_PATH}/macsec_{self.mymac.replace(':', '')}.key")[0],
+        )
+
+        # Detect if the server IP is IPv4 or IPv6 and create a socket accordingly
+        if ":" in self.sslServerIP:
+            clientSocket = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
+        else:
+            clientSocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+        # Make the client socket suitable for secure communication
+        self.secure_client_socket = context.wrap_socket(clientSocket)
+        try:
+            result = self.connection(self.secure_client_socket)
+            if result['authenticated']:
+                self.mua.auth_pass(secure_client_socket=self.secure_client_socket, client_mac=self.server_mac)
+            else:
+                self.mua.auth_fail(client_mac=self.server_mac)
+        except Exception as e:
+            self.logger.error("Define better this exception.", exc_info=True)
+            self.mua.auth_fail(client_mac=self.server_mac)
+        # finally:
+        #     # Close the socket
+        #     secureClientSocket.close()
+
+    def connection(self, secureClientSocket):
+        result = {
+            'IP': self.sslServerIP,
+            'authenticated': False
+        }
+
+        try:
+            self.to_validate(secureClientSocket, result)
+        except Exception as e:
+            self.logger.error("An error occurred during the connection process.", exc_info=True)
+
+        finally:
+            return result
+
+    def to_validate(self, secureClientSocket, result):
+        # If the IP is a link-local IPv6 address, connect it with the interface index
+        retries = 0
+        while retries < MAX_RETRIES:
+            try:
+                if self.sslServerIP.startswith("fe80"):
+                    secureClientSocket.connect(
+                        (self.sslServerIP, self.sslServerPort, 0, socket.if_nametoindex(self.interface)))
+                else:
+                    secureClientSocket.connect((self.sslServerIP, self.sslServerPort))
+                break  # break out of loop if connection is successful
+            except ConnectionRefusedError:
+                retries += 1
```

Comment:
Possible infinite loop: `retries` gets incremented only in the case of ConnectionRefusedError but the try block can fail for a variety of reasons like TimeoutError, ConnectionAbortedError, ConnectionResetError... I guess here you should use `except Exception as e` like you did above🙂